server/index.js

```diff
@@ -1,104 +1,164 @@
-const redirects = requireWithFallback('./routes/redirects') // AdHoc: Changed to requireWithFallback
-const errorPages = requireWithFallback('./routes/errors') // AdHoc: Changed to requireWithFallback
-
-const userAuth = requireWithFallback('userAuth')
-const customCsp = requireWithFallback('csp')
-
-const app = express()
-// AdHoc: Added Body Parser to support posts
-app.use(express.json())
-
-const {preload, postload} = allMiddleware
-
-// The trust proxy flag tells the app to use https for links
-// and redirect urls if it sees indications that the request
-// passed through a proxy and was originally sent using https
-if ((process.env.TRUST_PROXY || '').toUpperCase() === 'TRUE') {
-  app.enable('trust proxy')
-}
+const log = require('./logger');
 
-app.set('view engine', 'ejs')
-app.set('views', [path.join(__dirname, '../custom/layouts'), path.join(__dirname, '../layouts')])
+(async () => {
+  const gcpMetadata = require('gcp-metadata');
+  const { SecretManagerServiceClient } = require('@google-cloud/secret-manager');
 
-app.get('/healthcheck', (req, res) => {
-  res.send('OK')
-})
+  // GCP Metadata is only available to the deployed instance.
+  // Running locally, you would still use the .env file.
+  const isDeployedInstance = await gcpMetadata.isAvailable();
 
-app.use(csp({directives: customCsp}))
-app.use(userAuth)
+  if (isDeployedInstance) {
+    log.info('App Engine deployed instance detected...');
 
-preload.forEach((middleware) => app.use(middleware))
+    // We're on GCP here.  Load up environment variables from 
+    // Secrets Manager
+    const projectId = await gcpMetadata.project('project-id');
 
-app.use(userInfo)
+    log.info(`Loading environment variables for project: ${projectId}`);
 
-// serve all files in the public folder
-app.use('/assets', express.static(path.join(__dirname, '../public')))
+    // List of environment variables stored in base64-encoded format, rather than raw strings
+    const base64Vars = ['GOOGLE_APPLICATION_JSON'];
 
-// strip trailing slashes from URLs
-app.get(/(.+)\/$/, (req, res, next) => {
-  res.redirect(req.params[0])
-})
+    const client = new SecretManagerServiceClient();
 
-app.get('/view-on-site/:docId', (req, res, next) => {
-  const {docId} = req.params
-  const doc = getMeta(docId)
+    const parent = `projects/${projectId}`;
 
-  if (!doc) return next(Error('Not found'))
+    const [secrets] = await client.listSecrets({
+      parent: parent,
+    });
 
-  res.redirect(doc.path)
-})
+    for (let secret of secrets) {
+      const path = secret.name.split('/');
+      const envVar = path[path.length - 1];
 
-// main pages
-app.use(readingHistory.middleware)
+      log.info(`Loading environment variable ${envVar}`);
 
-// don't cache pages client-side to ensure browser always gets latest revision
-app.use((req, res, next) => {
-  res.set('Cache-Control', 'no-cache')
-  next()
-})
+      process.env[envVar] = await getSecret(client, secret.name, base64Vars.indexOf(envVar) !== -1);
```

Comment:
oh hey I think you can do the base64 thing automatically in code, too

Alec's handling added a label to the base64-encoded secret, see:
![image](https://user-images.githubusercontent.com/769353/181383526-5fe64109-d5cb-4b6a-b643-ee7486a24c2c.png)

and it looks like `listSecrets` returns `labels` (map of `string:string`) as part of the `Secret` object response:
https://cloud.google.com/secret-manager/docs/reference/rest/v1/projects.secrets#Secret

so I think it could be just something like this:

```suggestion
      process.env[envVar] = await getSecret(client, secret.name, secret.labels['encoded'] == 'base64');
```

Reply:
Oooh, good to know!